File: headapps/app/src/lib/factories/vehicle-factory.ts
import { Vehicle } from "../types/vehicle";
import { v4 as uuid } from "uuid";

/**
 * PATRÓN FACTORY METHOD (versión TypeScript).
 *
 * Mismos conocimientos aplicados al core migrado a Next.js: cada fábrica concreta decide
 * qué modelo crear, y la clase base aplica los valores por defecto (año actual) mediante
 * un paso tipo Builder. Agregar un modelo nuevo = una nueva subclase, sin tocar lo demás.
 */
export abstract class VehicleFactory {
  /** Factory Method: la subclase decide marca/modelo. */
  protected abstract createVehicle(): Pick<Vehicle, "brand" | "model">;

  /** Crea el vehículo completo con id y propiedades por defecto (año actual). */
  build(): Vehicle {
    const base = this.createVehicle();
    return {
      id: uuid(),
      brand: base.brand,
      model: base.model,
      year: new Date().getFullYear(), // año actual por defecto
    };
  }
}

export class MustangFactory extends VehicleFactory {
  protected createVehicle() {
    return { brand: "Ford", model: "Mustang" };
  }
}

export class ExplorerFactory extends VehicleFactory {
  protected createVehicle() {
    return { brand: "Ford", model: "Explorer" };
  }
}

export class EscapeFactory extends VehicleFactory {
  protected createVehicle() {
    return { brand: "Ford", model: "Escape" };
  }
}

File: DesignPatterns/Patterns/Factories/VehicleFactory.cs
using DesignPatterns.Models;
using DesignPatterns.Patterns.Builders;

namespace DesignPatterns.Patterns.Factories
{
    /// <summary>
    /// PATRÓN FACTORY METHOD (Método de Fábrica).
    ///
    /// Define el "molde" para crear vehículos, pero delega en cada subclase la decisión de
    /// QUÉ vehículo concreto se crea (CreateVehicle). El Arquitecto de Software prevé que
    /// el negocio agregará más modelos; con este patrón, agregar un modelo nuevo = crear una
    /// nueva fábrica concreta, SIN modificar el código existente (Abierto/Cerrado - SOLID).
    /// </summary>
    public abstract class VehicleFactory
    {
        /// <summary>
        /// Factory Method: cada fábrica concreta decide qué modelo de vehículo instanciar.
        /// </summary>
        protected abstract Vehicle CreateVehicle();

        /// <summary>
        /// Operación de fábrica de alto nivel: crea el vehículo y le aplica las
        /// propiedades por defecto usando el patrón Builder (año actual + propiedades del
        /// negocio). Así combinamos Factory Method (qué modelo) + Builder (cómo se configura).
        /// </summary>
        public Vehicle Build()
        {
            var vehicle = CreateVehicle();

            return new VehicleBuilder(vehicle)
                .WithCurrentYear()
                .WithDefaultProperties()
                .Build();
        }
    }
}
